lib/services/role.go

```diff
@@ -2748,6 +2752,102 @@ func (set RoleSet) checkAccess(r AccessCheckable, state AccessState, matchers ..
 		r.GetKind(), additionalDeniedMessage)
 }
 
+// checkRoleLabelsMatch checks if the [role] matches the labels of [resource]
+// for [condition].
+// It considers both the role labels (<kind>_labels) and label expression
+// (<kind>_labels_expression).
+//
+// Returns a match boolean, a debug message, and any unexpected error.
+//
+// If [condition] is types.Deny, the match is greedy, if either one matches it's
+// considered a match.
+//
+// If [condition] is types.Allow, the match is not greedy, if either doesn't
+// match it's not considered a match.
+//
+// If neither is set, it's not a match in either case.
+func checkRoleLabelsMatch(
+	condition types.RoleConditionType,
+	role types.Role,
+	resource AccessCheckable) (bool, string, error) {
+
+	labelMatchers, err := role.GetLabelMatchers(condition, resource.GetKind())
+	if err != nil {
+		return false, "", trace.Wrap(err)
+	}
+	return checkLabelsMatch(condition, labelMatchers, resource)
+}
+
+// checkLabelsMatch checks if the [labelMatchers] match the labels of [resource]
+// for [condition].
+// It considers both [labelMatchers.Labels] and [labelMatchers.Expression].
+//
+// Returns a match boolean, a debug message, and any unexpected error.
+//
+// If [condition] is types.Deny, the match is greedy, if either one matches it's
+// considered a match.
+//
+// If [condition] is types.Allow, the match is not greedy, if either doesn't
+// match it's not considered a match.
+//
+// If neither is set, it's not a match in either case.
+func checkLabelsMatch(
+	condition types.RoleConditionType,
+	labelMatchers types.LabelMatchers,
+	resource LabelGetter,
+) (bool, string, error) {
+	if labelMatchers.Empty() {
+		return false, "no label matchers or label expression", nil
+	}
+
+	var matches []bool
+	var messages []string
+
+	if len(labelMatchers.Labels) > 0 {
+		match, message, err := MatchLabelGetter(labelMatchers.Labels, resource)
+		if err != nil {
+			return false, "", trace.Wrap(err)
+		}
+		matches = append(matches, match)
+		messages = append(messages, "labels: "+message)
+	}
+
+	if len(labelMatchers.Expression) > 0 {
+		match, message, err := matchLabelExpression(labelMatchers.Expression, resource)
+		if err != nil {
+			return false, "", trace.Wrap(err)
+		}
+		matches = append(matches, match)
+		messages = append(messages, "expression: "+message)
+	}
+
+	message := strings.Join(messages, " ")
+
+	// Deny rules are greedy, if either matched, it's a match.
+	if condition == types.Deny {
+		return slices.Contains(matches, true), message, nil
```

Comment:
nit: wondering how the message looks like. for example:
```
labels: no match, empty selector expression: match
```
It's still readable but probably could improve

Reply:
note: we'll never get the "empty selector" message anymore because we check for empty in this function, it will either be `no key match: 'env'` or `no value match: got 'prod' want: '[staging]'`

Made some minor improvements, the message will now look like `windows_desktop_service \"\" denied, deny rule in role \"rolename\" matched; match(namespace=matched, label=no value match: got 'prod' want: '[staging]', expression=matched)"`